chainer/training/trainer.py

```diff
@@ -0,0 +1,285 @@
+        self.priority = priority
+
+
+class Trainer(object):
+
+    """The standard training loop in Chainer.
+
+    Trainer is an implementation of a training loop. Users can invoke the
+    training by calling the :meth:`run` method.
+
+    Each iteration of the training loop proceeds as follows.
+
+    - Update of the parameters. It includes the mini-batch loading, forward
+      and backward computations, and an execution of the update formula.
+      These are all done by the update object held by the trainer.
+    - Invocation of trainer extensions in the descending order of their
+      priorities. A trigger object is attached to each extension, and it
+      descides at each iteration whether the extension should be executed.
+      Trigger objects are callable objects that take the trainer object as the
+      argument and return a boolean value indicating whether the extension
+      should be called or not.
+
+    Extensions are callable objects that take the trainer object as the
+    argument. There are two ways to define custom extensions: inheriting the
+    :class:`Extension` class, and decorating functions by
+    :func:`make_extension`. See :class:`Extension` for more details on custom
+    extensions.
+
+    Users can register extensions to the trainer by calling the :meth:`extend`
+    method, where some configurations can be added.
+
+    - Trigger of an extension is a callable object that takes the trainer
+      object as the argument, and is called at each iteration. It decides
+      whether the extension should be executed at the iteration by returning a
+      boolean value. In most cases, :class:`IntervalTrigger` is used, in which
+      case users can simply specify a tuple of the interval length and its
+      unit, like ``(1000, 'iteration')`` or ``(1, 'epoch')``.
+    - The order of execution of extensions is determined by their priorities.
+      Extensions of higher priorities are invoked earlier. There are three
+      standard values for the priorities:
+
+      - ``PRIORITY_WRITER``. This is the priority for extensions that write
+        some records to the :attr:`observation` dictionary. It includes cases
+        that the extension directly adds values to the observation dictionary,
+        or the extension uses the :func:`chainer.report` function to report
+        values to the observation dictionary.
+      - ``PRIORITY_EDITOR``. This is the priority for extensions that edit the
+        :attr:`observation` dictionary based on already reported values.
+      - ``PRIORITY_READER``. This is the priority for extensions that only read
+        records from the :attr:`observation` dictionary. This is also suitable
+        for extensions that do not use the :attr:`observation` dictionary at
+        all.
+
+    - Extensions with ``invoke_before_training`` flag on are also invoked at
+      the beginning of the training loop. Extensions that update the training
```

Comment:
Yes, these extensions are invoked before training begins as well as at each iteration (when the trigger fires).